_Main topics covered during today's session:_

This NB:

1. **NBA Data sample exercises**

Next NB:

1. **Higher order functions**
2. **Slicing**
3. **2 Ways of computing a solution (plus a bonus method for take home)**

In [1]:
!wget https://raw.githubusercontent.com/gt-cse-6040/skills_oh_week_04/main/nba_json.txt

--2023-09-19 23:48:56--  https://raw.githubusercontent.com/gt-cse-6040/skills_oh_week_04/main/nba_json.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8903 (8.7K) [text/plain]
Saving to: ‘nba_json.txt’

nba_json.txt        100%[===================>]   8.69K  --.-KB/s    in 0s      

2023-09-19 23:48:56 (49.3 MB/s) - ‘nba_json.txt’ saved [8903/8903]



## Nested Data Structures: NBA Player Analysis -- Part 2

### This notebook is an application of the NBA nested data notebook.

#### We will be taking that data structure and performing some typical tasks that we do on dictionaries.

In [2]:
# loading the data, same as the previous NB
# import the required package
import ast

# reading the data from the json file
with open('nba_json.txt') as f:
    data = f.readlines()

nested = [] # this is going to be our nested data structure
#put each line(string) into a dictionary
for line in data:
    result = ast.literal_eval(line)
#     print(result)
    nested.append(result)

### As we know from the previous NB, the key 'resultSets' contains the data the we want to manipulate.

### Let's take a look at it again.

In [3]:
print(nested[0]['resultSets'])
print('\n')
print(nested[0]['resultSets'][0].keys())

[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1,

## Note that this is a list containing a single dictionary, with three keys.

#### The key 'rowSet' contains the statistics that we are interested in, and the key 'headers' defines what each of the statisics is.

#### Note that this first dict is for the player Al-Farouq Aminu, with the player_id 202329.

In [4]:
print(nested[0]['resultSets'][0]['headers'])
print('\n')
print(nested[0]['resultSets'][0]['rowSet'])
print('\n')
print(nested[0]['resultSets'][0]['rowSet'][0])
print('\n')
print(nested[0]['resultSets'][0]['rowSet'][1])

['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC']


[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 0.0, 0, 0, 0, 0, 0, 0, 0, 316.8]]


['22019', 1

## For simplicity, we are going to work with the nested[0]['resultSets'][0] list, which gives us a dictionary having three keys, as we saw above.

### This will allow us to address the 'rowSet' dictionary key for its values.

## Exercise 0

### Who are the players that a player guarded during the season?

#### In other words, what are the OFF_PLAYER_NAME's for the DEF_PLAYER_NAME?

### Requirement:  Output a list of the names of the players he guarded.


In [5]:
def players_guarded(var_stats, DEF_PLAYER_NAME):

    # create the list to return
    ret_list = []
    print(var_stats)
#     print('\n')
#     print(DEF_PLAYER_NAME)

    for k,v in var_stats.items():
        if k == 'rowSet':
            # remember that v is a list of lists
#             print(v[0])
            for index in v:  # loops over the higher level list
#                 print(index[2])
                # index is the list within the higher level list
                ret_list.append(index[2])

    return ret_list

# test code
list0 = players_guarded(nested[0]['resultSets'][0], 'Al-Farouq Aminu')
print(list0)
list0 = players_guarded(nested[1]['resultSets'][0], 'Bam Adebayo')
print(list0)
list0 = players_guarded(nested[6]['resultSets'][0], 'Giannis Antetokounmpo')
print(list0)

{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 

## Exercise 1

### Now let's make this a little more useful, asking the same question of who are the players guarded during the season.

### Requirement:  Output a dictionary with the key being the defensive player name and the value being a list of the offensive players he guarded.

In [ ]:
def players_guarded_dict(var_stats, DEF_PLAYER_NAME):

# Same as above************************
    # create the list
    guarded_list = []
#     print(var_stats)

    # loop over the statistics dict
    # this is the same as the above
    for k,v in var_stats.items():
        if k == 'rowSet':
            # remember that v is a list of lists
#             print(v[0])
            for index in v:  # loops over the higher level list
        #         print(index[2])
                # index is the list within the higher level list
                guarded_list.append(index[2])
# *************************************

# This is new**************************
    # use default dict
    from collections import defaultdict
    ret_dict = defaultdict()
    # now put it all into the default dictionary
    ret_dict[DEF_PLAYER_NAME] = guarded_list

    # make the default dict a regular dict and return
    return dict(ret_dict)
# **************************************

# test code
list1 = players_guarded_dict(nested[0]['resultSets'][0], 'Al-Farouq Aminu')
print(list1)
list1 = players_guarded_dict(nested[1]['resultSets'][0], 'Bam Adebayo')
print(list1)
list1 = players_guarded_dict(nested[6]['resultSets'][0], 'Giannis Antetokounmpo')
print(list1)

## Exercise 2

### Again, let's make this a little more useful, asking the same question of who are the players guarded during the season. We want to know how many points he allowed to each player during the season.

#### Note that this data element is called 'PLAYER_PTS'. Which data element in the list is it?

### Requirement:  Output a dictionary with the key being the defensive player name and the value being a list of tuples. The first item of each tuple is the name of the offensive players he guarded, and the second item of the tuple is the number points allowed.

In [ ]:
def players_guarded_minutes_dict(var_stats, DEF_PLAYER_NAME):

# Same as above***********************
    # create the list
    guarded_list = []
#     print(var_stats)
# ************************************


# This is new*************************

    # loop over the statistics dict
    # this is the same as the above
    for k,v in var_stats.items():
        if k == 'rowSet':
            # remember that v is a list of lists
#             print(v[0])
            for index in v:  # loops over the higher level list
        #         print(index[2])
                # index is the list within the higher level list
            # This is the new part********************
                guarded_list.append((index[2],index[8]))
            # ****************************************

# ************************************

# Same as above************************
    # use default dict
    from collections import defaultdict
    ret_dict = defaultdict()
    # now put it all into the default dictionary
    ret_dict[DEF_PLAYER_NAME] = guarded_list

    # make the default dict a regular dict and return
    return dict(ret_dict)
# *************************************

# test code
list2 = players_guarded_minutes_dict(nested[0]['resultSets'][0], 'Al-Farouq Aminu')
print(list2)
list2 = players_guarded_minutes_dict(nested[1]['resultSets'][0], 'Bam Adebayo')
print(list2)
list2 = players_guarded_minutes_dict(nested[6]['resultSets'][0], 'Giannis Antetokounmpo')
print(list2)

## Exercise 3

### Again, let's make this a little more useful, asking the same question of who are the players guarded during the season. We want to know how many pointed he allowed to each player during the season, and who scored the least to most..

### Requirement:  Output a dictionary with the key being the defensive player name and the value being a list of tuples. The first item of each tuple is the name of the offensive players he guarded, and the second item of the tuple is the number of points allowed. The list of tuples should be sorted by the number of points allowed, from most to least. If the points are the same, then sort by the name in alphabetical order.

In [ ]:
def players_guarded_minutes_dict_sorted(var_stats, DEF_PLAYER_NAME):

# Same as above***********************
    # create the list
    guarded_list = []
#     print(var_stats)

    # loop over the statistics dict
    # this is the same as the above
    for k,v in var_stats.items():
        if k == 'rowSet':
            # remember that v is a list of lists
#             print(v[0])
            for index in v:  # loops over the higher level list
                guarded_list.append((index[2],index[8]))

# ************************************

# Sort the list of tuples, this is new*******
# https://www.geeksforgeeks.org/python-mutiple-column-sort-in-tuples/
# https://www.geeksforgeeks.org/python-program-to-sort-a-list-of-tuples-by-second-item/
    guarded_list = sorted(guarded_list, key = lambda x: (-x[1], x[0]))


# *******************************************

# Same as above************************
    # use default dict
    from collections import defaultdict
    ret_dict = defaultdict()
    # now put it all into the default dictionary
    ret_dict[DEF_PLAYER_NAME] = guarded_list

    # make the default dict a regular dict and return
    return dict(ret_dict)
# *************************************

# test code
list2 = players_guarded_minutes_dict_sorted(nested[0]['resultSets'][0], 'Al-Farouq Aminu')
print(list2)
list2 = players_guarded_minutes_dict_sorted(nested[1]['resultSets'][0], 'Bam Adebayo')
print(list2)
list2 = players_guarded_minutes_dict_sorted(nested[6]['resultSets'][0], 'Giannis Antetokounmpo')
print(list2)